src/scheduler/resv_info.c

```diff
@@ -1636,92 +1656,148 @@ confirm_reservation(status *policy, int pbs_sd, resource_resv *unconf_resv, serv
+ * @retval	-3	: error
  */
 int
-check_vnodes_down(resource_resv *resv, int *tot_vnodes, char *names_of_down_vnodes)
+check_vnodes_unavailable(resource_resv *resv)
 {
-	int nn_length; /* number of characters allowable in log buffer */
-	int vnodes_down = 0; /* number of unavailable vnodes */
-	int j;
-	int tot_chk_cnt;
-	int chk_ind = 0;
-	chunk *chk;
-
-	if (resv == NULL || resv->nspec_arr == NULL || tot_vnodes == NULL)
-		return -2;
-
-	*tot_vnodes = 0;
-	chk = resv->select->chunks[0];
-	/* check chunk in the select spec has the number of chunks of that type we need.
-	 * tot_chnk_cnt is the number of chunks of the previous chunks plus the number of this chunk.
-	 * Once we've reached this number of nodes we know we have to move onto the next chunk.
-	 */
-	tot_chk_cnt = chk->num_chunks;
+	int ret = 0;
+	int i;
+	int has_down_node = 0;
+	int has_superchunk = 0;
+	int num_chunks = 0;
+	nspec **chunks_to_remove = NULL;
+	int del_i = 0;
+
+	if (resv == NULL || resv->nspec_arr == NULL)
+		return -3;
+
+	for (i = 0; resv->orig_nspec_arr[i] != NULL; i++) {
+		if (!resv->orig_nspec_arr[i]->end_of_chunk)
+			has_superchunk = 1;
+		num_chunks++;
+	}
 
-	for (j = 0; resv->nspec_arr[j] != NULL; j++) {
-		node_info *ninfo = resv->nspec_arr[j]->ninfo;
-		(*tot_vnodes)++;
+	if (has_superchunk) {
+		if ((chunks_to_remove = malloc((num_chunks + 1) * sizeof(nspec *))) == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
+			return -3;
+		}
+	}
 
-		if (j == tot_chk_cnt) {
-			chk_ind++;
-			chk = resv->select->chunks[chk_ind];
-			tot_chk_cnt += chk->num_chunks;
+	for (i = 0; resv->orig_nspec_arr[i] != NULL; i++) {
+
+		/* If the original select chunks haven't been mapped to the resv_nodes and the reservation is running, we can't continue */
+		if (resv->resv->resv_state == RESV_RUNNING && resv->orig_nspec_arr[i]->chk == NULL) {
+			free(chunks_to_remove);
+			return -2;
 		}
 
-		if (ninfo->is_down || ninfo->is_offline || ninfo->is_stale || ninfo->is_unknown) {
-			int k, m;
-			/* We can't reconfirm a reservation if there is a running job on one of our unavailable nodes */
-			if (resv->resv->resv_queue->running_jobs != NULL)
-				for (k = 0; resv->resv->resv_queue->running_jobs[k] != NULL; k++) {
-					resource_resv *job = resv->resv->resv_queue->running_jobs[k];
-					for (m = 0; job->ninfo_arr[m] != NULL; m++)
-						if (job->ninfo_arr[m]->rank == ninfo->rank)
-							return -1;
-				}
+		/* Part of a superchunk we've already handled */
+		if (resv->orig_nspec_arr[i]->ninfo == NULL)
+			continue;
 
-			vnodes_down++;
-			resv->nspec_arr[j]->ninfo = NULL;
+		ret = check_down_running(resv, i);
+		/* running jobs on unavailable nodes in chunk */
+		if (ret == -2) {
+			free(chunks_to_remove);
+			return -1;
+		}
+		/* unavailable nodes in chunk */
+		else if (ret == -1) {
+			int j;
+			has_down_node = 1;
+			for (j = i; resv->orig_nspec_arr[j] != NULL && !resv->orig_nspec_arr[j]->end_of_chunk; j++) {
+				resv->orig_nspec_arr[j]->ninfo = NULL;
```

Comment:
Don't we need to free resreq and reassign it as we do at line 1786, 1787?

Reply:
No, these are the chunks we will delete.  When we detect a superchunk with a down node, we delete all chunks except the last one.  We then have it request all the resources for the superchunk so it will be allocated elsewhere.  It doesn't matter if any node of the superchunk is up, as long as there is one down, the whole thing gets reassigned.